File: python/simplify_path.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        """time complexity: O(n)"""
        stack: list[str] = []
        path += "/"
        curr: str = ""
        for c in path:
            if c == "/":
                if curr == ".." and len(stack) > 0:
                    stack.pop()
                if curr != "." and curr != ".." and curr != "":
                    stack.append(curr)
                curr = ""
            else:
                curr += c
        return "/" + "/".join(stack)


def test_simplifyPath_case_1():
    # arrange
    path: str = "/home/"
    expected: str = "/home"

    # act
    solution = Solution()
    actual = solution.simplifyPath(path)

    # assert
    assert expected == actual


def test_simplifyPath_case_2():
    # arrange
    path: str = "/../"
    expected: str = "/"

    # act
    solution = Solution()
    actual = solution.simplifyPath(path)

    # assert
    assert expected == actual


def test_simplifyPath_case_3():
    # arrange
    path: str = "/home//foo/"
    expected: str = "/home/foo"

    # act
    solution = Solution()
    actual = solution.simplifyPath(path)

    # assert
    assert expected == actual


def test_simplifyPath_case_4():
    # arrange
    path: str = "/home/user/Documents/../Pictures"
    expected: str = "/home/user/Pictures"

    # act
    solution = Solution()
    actual = solution.simplifyPath(path)

    # assert
    assert expected == actual

File: rust/src/code/simplify_path.rs
struct Solution {}

impl Solution {
    pub fn simplify_path(path: String) -> String {
        let mut path: Vec<char> = path.chars().collect();
        path.push('/');
        let mut stack: Vec<String> = Vec::new();
        let mut curr: Vec<char> = Vec::new();
        for i in 0..path.len() {
            if path[i] == '/' {
                let curr_string: String = curr.iter().collect::<String>();
                if curr_string == "..".to_string() {
                    if stack.len() > 0 {
                        stack.pop().unwrap();
                    }
                } else if curr_string != ".".to_string() && !curr.is_empty() {
                    stack.push(curr_string);
                }
                curr.clear();
            } else {
                curr.push(path[i]);
            }
        }

        "/".to_string() + &stack[..].join("/")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn daily_simplify_path_case_1() {
        // arrange
        let path: String = "/home/".to_string();
        let expected: String = "/home".to_string();

        // act
        let actual = Solution::simplify_path(path);

        assert_eq!(expected, actual);
    }

    #[test]
    fn daily_simplify_path_case_2() {
        // arrange
        let path: String = "/../".to_string();
        let expected: String = "/".to_string();

        // act
        let actual = Solution::simplify_path(path);

        assert_eq!(expected, actual);
    }

    #[test]
    fn daily_simplify_path_case_3() {
        // arrange
        let path: String = "/home/user/Documents/../Pictures".to_string();
        let expected: String = "/home/user/Pictures".to_string();

        // act
        let actual = Solution::simplify_path(path);

        assert_eq!(expected, actual);
    }
}
